### DISCUSSION TITLE
What do these HyperHDR performance statistics mean?

### DISCUSSION THREAD

**Discussion by bruvv** [awawa-dev/HyperHDR#479] [Q&A] [answered]
<img width="696" alt="image" src="https://user-images.githubusercontent.com/3063928/211154097-4455825c-058a-4e3a-abe5-81a280d75ff6.png">

**Comment by awawa-dev** [comment, marked as answer]
Arrows up/down bext to the number points the stream directions so:
USB grabber: captured~600 frames and forwarded to HyperHDR instance ~600 frames.
Instance: received and processed (including image to LEDs colors processing) ~600 frames. On the output (also ~600) is not the image anymore but LED colors.
You have a smoothing enabled so the smoothing received 600 LEDs strip colors and forwarded 5000 final smoothed LEDs strip colors to the LED device (not on the screenshot but can be read from it).
LED device received 5000 LED strip colors and successfully wrote all of them (5000) to the physical device.

**Reply by bruvv** [reply]
Thanks! And knowing this, is there a way to optimize it more? I am missing a bit of optimization documentation :)
This is what my grabber is auto set to: `1280x720x30 MJPEG` and this is the rp 3b data:
<img width="560" alt="image" src="https://user-images.githubusercontent.com/3063928/211155050-c0e57818-0397-4c73-9291-686491882b67.png">
(not sure why there is undervolting, running 216 SK6812 leds with a 5v 10A power adapter)
And this is the smoothing:
<img width="664" alt="image" src="https://user-images.githubusercontent.com/3063928/211155088-0f810b9e-c73f-4a2d-9da5-dac718b53330.png">

How can I optimize it more running 1080p non HDR content.

In the logs I see a lot of:
`2023-01-07T14:13:50.575Z [IMAGETOLED0] This is large image area for lamp: 207. It contains 1892 indexes for captured video frame so reduce it by four. Enabling 'sparse processing' option for you. Consider to enable it permanently in the processing configuration to hide that warning.`

But as english is not my first language I fail to understand what: `This is large image area for lamp` means

**Reply by awawa-dev** [reply]
1 'undervoltage' message: read second point https://github.com/awawa-dev/HyperHDR/discussions/320 . It's a serious issue.
2 `In the logs I see a lot of:` it's just a warning. HyperHDR automatically chooses best performance approach when converting the image to the LED strips colors (e.g. for a LED index 207 there are almost 1892 pixels to compute a single LED color...your video resolution is too high, still HyperHDR and your Rpi is strong enough to process 720p stream with only 8ms lag).
3 not sure why with 1280x720x30 MJPEG you have only 10FPS but other settings in the grabber settings or the under-voltage problem could affect it. Full HyperHDR logs taken after 2-3 minutes from the start are needed.

**Reply by bruvv** [reply]
I do not understand how it can be under-volting? I have a overpowered power supply, should have around 4A available for the pi. Can you elaborate this a bit more?

And I just upgrade my grabber this is the new data:
<img width="333" alt="image" src="https://user-images.githubusercontent.com/3063928/211155882-1c8f870f-8b27-4dc5-adb4-0f976d395c1f.png">

What would you say I can optimize? Can I update my update frequency for example?

**Reply by awawa-dev** [reply]
`undervoltage` is reported by the Raspberry Pi OS and HyperHDR just reads it from the OS logs. Direct all complaints to your Rpi sensors if you believe your power supply is strong and stable enough (but it isn't in fact...) 😉  When LED strip power supply is used for Rpi, almost always it is not about amps, but about stability.

**Reply by awawa-dev** [reply]
For you current video grabber settings enable 'Quarter of frame mode' in the grabber settings. It should improve current 40ms lag.

**Reply by bruvv** [reply]
That is an awesome setting it helps indeed! And what about my led update frequency, any tuning there possible?

**Reply by awawa-dev** [reply]
Dont know why you set non-standard very high 450ms delay (150ms is default) in the smoothing. 
Usually 80Hz is fine for the refresh rate but read the real statistics from the device: https://github.com/awawa-dev/HyperHDR/wiki/HyperSerial

**Reply by awawa-dev** [reply]
General rule for setting USB grabber is to have captured video stream at lowest resolution (to reduce processing time, still should be fine for ambient lightning) and highest rate (to reduce lag at the input from USB grabber). We really don't need capture video in 1080p or 720p and lower settings should be used but not every USB grabber has a built-in scaler so 'Quarter of frame mode' can be used as workaround.

**Reply by bruvv** [reply]
Thanks for the explanations, I have adjusted my settings as followed:
`720x480x60 MJPEG`
<img width="663" alt="image" src="https://user-images.githubusercontent.com/3063928/211161664-98f5574d-a39e-4994-84f1-4c50d9fe2709.png">
Why I have my smoothing time higher? I find it a bit relaxing when it smooths over in colors instead of hard cutting, which is what I expierence with 150ms time.

**Reply by awawa-dev** [reply]
Sure, the most important thing is to adjust the settings to your own requirements, I just wanted to be sure you are aware of the consequences in the form of an increased lag. 

If you have an alternative for `720x480x60 MJPEG` using NV12 or YUV codec (like e.g. `720x480x60 NV12`) choose them instead. MJPEG requires a lot more of resources to decode it.

**Reply by awawa-dev** [reply]
On the other hand, you're using a slow USB 2.0, which unfortunately favors compressed MJPEG (BTW it may have cause problems with HyperSerialESP32 described on the HyperSPI project as well). You must do some tests and assess which grabber setup produce lower lag and has lower impact on the CPU.

**Reply by bruvv** [reply]
Good point, the only thing is, if I want 60 FPS with YUV there is no option:
<img width="1131" alt="image" src="https://user-images.githubusercontent.com/3063928/211163603-05151e90-de00-42db-8c66-bdc58de7a930.png">
but with 30:
<img width="1127" alt="image" src="https://user-images.githubusercontent.com/3063928/211163613-fcfc8a68-56f1-4f2a-9978-4789b43e9327.png">
What would you recommend?

**Reply by awawa-dev** [reply]
Don't you have nv12 (on ms2109 I supposed, maybe nv12@60 works only on Windows)? If not then with MJPEG@60.

**Reply by bruvv** [reply]
I do not have nv12 with [this](https://www.amazon.com/DIGITNOW-Converter-Recording-Streaming-Broadcasting/dp/B097DBWKFT/ref=sr_1_19?crid=KC2N6B4P1LOZ&keywords=hdmi+grabber&sprefix=hdmi+grabbe%2Caps%2C161&sr=8-19) grabber. Running on a pi 3b. I will switch to MJPEG@60, any tips on resolution?
